src/index.ts

```diff
@@ -534,6 +571,177 @@ function printStatus(e: Event) {
+    if (evt.data instanceof Buffer || evt.data instanceof ArrayBuffer) {
+      const buf = Buffer.from(evt.data);
+      try {
+        EmptyReply.decode(buf);
+        prependLog(`got expected empty reply`);
+      } catch (e) {
+        prependLog(`error: ${e}`);
+      }
+    }
+  };
+
+  ws.onerror = (evt: { error: Error }) => {
+    console.log(evt);
+    prependLog(`error: ${evt.error}`);
+  };
+
+  ws.onclose = (evt: { code: number; reason: string }) => {
+    if (evt.code === 1006) {
+      prependLog(`wrong password`);
+      return;
+    }
+
+    prependLog(`closed: ${evt.code} ${evt.reason}`);
+  };
+
+  const bytes = Buffer.from(Follow.$type.encode(msg).finish());
+  ws.onopen = () => {
+    ws.send(bytes);
+  };
+}
+
+function sqlStopFollow(e: Event) {
+  var password = adminPasswordInput.value;
+
+  const ws = new WebSocket(`${byprosURL}/Unfollow/${password}`);
+  ws.binaryType = "arraybuffer";
+
+  const msg = new Unfollow();
+
+  ws.onmessage = (evt: { data: WebSocket.Data }): any => {
+    if (evt.data instanceof Buffer || evt.data instanceof ArrayBuffer) {
+      const buf = Buffer.from(evt.data);
+      try {
+        EmptyReply.decode(buf);
+        prependLog(`got expected empty reply`);
+      } catch (e) {
+        prependLog(`error: ${e}`);
+      }
+    }
+  };
+
+  ws.onerror = (evt: { error: Error }) => {
+    console.log(evt);
+    prependLog(`error: ${evt.error}`);
+  };
+
+  ws.onclose = (evt: { code: number; reason: string }) => {
+    if (evt.code === 1006) {
+      prependLog(`wrong password`);
+      return;
+    }
+
+    prependLog(`closed: ${evt.code} ${evt.reason}`);
+  };
+
+  const bytes = Buffer.from(Unfollow.$type.encode(msg).finish());
+  ws.onopen = () => {
+    ws.send(bytes);
+  };
+}
+
+function sqlCatchup(e: Event) {
+  var password = adminPasswordInput.value;
+
+  const ws = new WebSocket(`${byprosURL}/CatchUpMsg/${password}`);
+  ws.binaryType = "arraybuffer";
+
+  const blockHex = catchupInput.value;
+  const block = hex2Bytes(blockHex);
+
+  const updateEvery = parseInt(catchupUpdateInput.value, 10);
+
+  console.log("catchup update input:", updateEvery);
+
+  const msg = new CatchUpMsg();
+  msg.fromblock = block;
+  msg.scid = hex2Bytes(
+    "9cc36071ccb902a1de7e0d21a2c176d73894b1cf88ae4cc2ba4c95cd76f474f3"
+  );
+  msg.target = roster.list[0];
+  msg.updateevery = updateEvery;
+
+  ws.onmessage = (evt: { data: WebSocket.Data }): any => {
+    if (evt.data instanceof Buffer || evt.data instanceof ArrayBuffer) {
+      const buf = Buffer.from(evt.data);
+      try {
+        const resp = CatchUpResponse.decode(buf);
+        prependLog(`catch up response: ${resp.toString()}`);
+      } catch (e) {
+        prependLog(`error: ${e}`);
+      }
+    }
+  };
+
+  ws.onerror = (evt: { error: Error }) => {
+    console.log(evt);
+    prependLog(`error: ${evt.error}`);
+  };
+
+  ws.onclose = (evt: { code: number; reason: string }) => {
+    if (evt.code === 1006) {
```

Comment:
Are we sure this will only occur for a wrong password? https://tools.ietf.org/html/rfc6455#section-7.4.1 suggests 1006 is reserved for abnormal close.